substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SubstitutionReportFeature.java

```diff
@@ -0,0 +1,188 @@
+                }
+            }
+        }
+    }
+
+    private void findSubstitutedMethods(FeatureImpl.AfterAnalysisAccessImpl access) {
+        AnalysisUniverse universe = access.getUniverse();
+        for (AnalysisMethod method : universe.getMethods()) {
+            if (method.wrapped instanceof SubstitutionMethod) {
+                SubstitutionMethod subMethod = (SubstitutionMethod) method.wrapped;
+                if (subMethod.isUserSubstitution()) {
+                    String jarLocation = getJarLocation(subMethod.getAnnotated().getDeclaringClass());
+                    substitutions.putIfAbsent(jarLocation, new Substitutions());
+                    substitutions.get(jarLocation).addMethod(subMethod);
+                }
+            }
+        }
+    }
+
+    private void findSubstitutedFields(FeatureImpl.AfterAnalysisAccessImpl access) {
+        AnalysisUniverse universe = access.getUniverse();
+        for (AnalysisField field : universe.getFields()) {
+            if (field.wrapped instanceof SubstitutionField) {
+                SubstitutionField subField = (SubstitutionField) field.wrapped;
+                if (subField.isUserSubstitution()) {
+                    String jarLocation = getJarLocation(subField.getAnnotated().getDeclaringClass());
+                    substitutions.putIfAbsent(jarLocation, new Substitutions());
+                    substitutions.get(jarLocation).addField(subField);
+                }
+            }
+        }
+    }
+
+    private void reportSubstitutions() {
+        ReportUtils.report("substitutions performed by native-image","reports", "substitutions", "csv", pw -> {
+            pw.println("location , category (type/method/field) , original , annotated");
+            for (Map.Entry<String, Substitutions> g : substitutions.entrySet()) {
+                for (Map.Entry<ResolvedJavaType, ResolvedJavaType> e : g.getValue().getSubstitutedTypes().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "type", e.getKey(), e.getValue(), t -> t.toJavaName(true)));
+                }
+                for (Map.Entry<ResolvedJavaMethod, ResolvedJavaMethod> e : g.getValue().getSubstitutedMethods().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "method", e.getKey(), e.getValue(), this::formatMethod));
+                }
+                for (Map.Entry<ResolvedJavaField, ResolvedJavaField> e : g.getValue().getSubstitutedFields().entrySet()) {
+                    pw.println(formatSubstitution(g.getKey(), "field", e.getKey(), e.getValue(), this::formatField));
+                }
+            }
+        });
+    }
+
+    private String getJarLocation(ResolvedJavaType type) {
+        Class<?> annotatedClass = OriginalClassProvider.getJavaClass(GraalAccess.getOriginalSnippetReflection(), type);
+        CodeSource source = annotatedClass.getProtectionDomain().getCodeSource();
+        return source == null ? "unknown" : source.getLocation().toString();
```

Comment:
When do we have unknown?

Reply:
`ProtectionDomain.getCodeSource` can return `null`, not specified under which conditions but after some research it seems to be `null` in the implementations of JDK classes. Since `getJarLocation` method is invoked only for subsitutee classes, there should be no issues, however I left this as a "failsafe".

Reply:
Is this specific to JARs or would it pick up locations in a folder on the classpath as well? In case it also looks at the folders on the classpath, it might be worth renaming it to, for e.g., `getTypeClassFileLocation`

Reply:
It isn't JAR specific, so it makes sense to rename it.